src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java

```diff
@@ -0,0 +1,219 @@
+				.withPayload("message1")
+				.setHeader(AwsHeaders.PARTITION_KEY, "fooKey")
+				.setHeader(AwsHeaders.SEQUENCE_NUMBER, "10")
+				.setHeader("foo", "bar")
+				.build();
+
+
+		ArgumentCaptor<UserRecord> userRecordRequestArgumentCaptor = ArgumentCaptor
+				.forClass(UserRecord.class);
+
+		this.kinesisSendChannel.send(message);
+		verify(this.kinesisProducer).addUserRecord(userRecordRequestArgumentCaptor.capture());
+		verify(this.kinesisProducer, Mockito.times(3)).getOutstandingRecordsCount();
+		UserRecord userRecord = userRecordRequestArgumentCaptor.getValue();
+		assertThat(userRecord.getStreamName()).isEqualTo("foo");
+		assertThat(userRecord.getPartitionKey()).isEqualTo("fooKey");
+		assertThat(userRecord.getExplicitHashKey()).isNull();
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	void testKPLMessageHandler_raw_payload_backpressure_failure_test() {
+		given(this.kinesisProducer.addUserRecord(any(UserRecord.class)))
+				.willReturn(mock(ListenableFuture.class));
+		this.kplMessageHandler.setMaxRecordsInFlight(2);
+		this.kplMessageHandler.setMaxInFlightRecordsBackoffDuration(100);
+		this.kplMessageHandler.setMaxInFlightRecordsBackoffMaxAttempts(2);
+		this.kplMessageHandler.setMaxInFlightRecordsBackoffRate(2);
+		given(this.kinesisProducer.getOutstandingRecordsCount())
+				.willReturn(5)
+				.willReturn(4)
+				.willReturn(3);
+		final Message<?> message = MessageBuilder
+				.withPayload("message1")
+				.setHeader(AwsHeaders.PARTITION_KEY, "fooKey")
+				.setHeader(AwsHeaders.SEQUENCE_NUMBER, "10")
+				.setHeader("foo", "bar")
+				.build();
+
+		ArgumentCaptor<UserRecord> userRecordRequestArgumentCaptor = ArgumentCaptor
+				.forClass(UserRecord.class);
+
+		try {
+			this.kinesisSendChannel.send(message);
+		}
+		catch (Exception ex) {
+			assertThat(ex).isNotNull();
+			assertThat(ex.getCause()).isNotNull();
+			assertThat(ex.getCause().getClass()).isEqualTo(RuntimeException.class);
+			assertThat(ex.getCause().getClass()).isEqualTo(RuntimeException.class);
+			assertThat(ex.getCause().getMessage()).isEqualTo("KPL Buffer already at max capacity.");
+		}
+
+		verify(this.kinesisProducer, Mockito.times(0)).addUserRecord(any(UserRecord.class));
+		verify(this.kinesisProducer, Mockito.times(3)).getOutstandingRecordsCount();
+	}
+
+	@AfterEach
+	public void tearDown() {
+		clearInvocations(this.kinesisProducer);
+	}
+
+	@Configuration
+	@EnableIntegration
+	public static class ContextConfiguration {
+
+		@Bean
+		public KinesisProducer kinesisProducer() {
+			return mock(KinesisProducer.class);
+		}
+
+		@Bean
+		@ServiceActivator(inputChannel = "kinesisSendChannel")
+		public MessageHandler kplMessageHandler(KinesisProducer kinesisProducer) {
+			KplMessageHandler kplMessageHandler = new KplMessageHandler(kinesisProducer);
+			kplMessageHandler.setAsync(true);
+			kplMessageHandler.setStream("foo");
+			kplMessageHandler.setEmbeddedHeadersMapper(new EmbeddedJsonHeadersMessageMapper("foo"));
```

Comment:
Initially there were no tests for KPL Message Handler. This could have been handy for the upcoming tests.  
Agreed, removed the references to keep only the contextual information.